PyRoute/DeltaReduce.py

```diff
@@ -0,0 +1,325 @@
+                    chunks[i] = []
+                    remove.append(i)
+                    best_sectors = temp_sectors
+                    msg = "Reduction found: new input has " + str(len(best_sectors.lines)) + " lines"
+                    self.logger.error(msg)
+
+            if 0 < len(remove):
+                num_chunks -= len(remove)
+
+            num_chunks *= 2
+
+            segment = best_sectors.lines
+            # if we're about to bust our loop condition, make sure we verify 1-minimality as our last hurrah
+            if num_chunks > len(segment) and not singleton_run:
+                singleton_run = True
+                num_chunks = len(segment)
+
+        # now that the pass is done, update self.sectors with best reduction found
+        self.sectors = best_sectors
+        if short_msg is not None:
+            self.logger.error("Shortest error message: " + short_msg)
+
+    def reduce_line_two_minimal(self):
+        segment = self.sectors.lines
+
+        # An interesting less-than-4-element list is 2-minimal by definition
+        if 4 > len(segment):
+            return
+
+        best_sectors = self.sectors
+        gap = 1
+        while gap < len(segment):
+            msg = "# of lines: " + str(len(best_sectors.lines))
+            self.logger.error(msg)
+
+            i = 0
+            j = i + gap
+            while j < len(segment):
+                lines_to_remove = [segment[i], segment[j]]
+                temp_sectors = best_sectors.drop_lines(lines_to_remove)
+
+                interesting, msg = self._check_interesting(self.args, temp_sectors)
+
+                # We've found a chunk of input and have _demonstrated_ its irrelevance,
+                # empty that chunk, update best so far, and continue
+                if interesting:
+                    best_sectors = temp_sectors
+                    segment = best_sectors.lines
+                    msg = "Reduction found: new input has " + str(len(best_sectors.lines)) + " lines"
+                    self.logger.error(msg)
+                else:
+                    i += 1
+                j = i + gap
+
+            gap += 1
+
+        # now that the pass is done, update self.sectors with best reduction found
+        self.sectors = best_sectors
+
+    def _assemble_all_but_ith_chunk(self, chunks, i):
+        # Assemble all _but_ the ith chunk
+        nulines = [item for ind, item in enumerate(chunks) if ind != i and ind < len(chunks)]
+        # pythonically flatten nulines (list of lists) into single list
+        raw_lines = [item for sublist in nulines for item in sublist]
+        return raw_lines
+
+    @staticmethod
+    def update_short_msg(msg, short_msg):
+        if msg is not None:
+            if short_msg is None:
+                short_msg = msg
+            elif len(msg) < len(short_msg):
+                short_msg = msg
+        return short_msg
+
+    @staticmethod
+    def _check_interesting(args, sectors):
```

Comment:
This runs the entire pathfinding/map-gen/stat-gen process and tracks what, if anything, goes boom.

Passing in a DeltaDictionary object (ie, sectors) mandates using DeltaGalaxy instead of plain galaxy, as the former has the gubbins needed to handle DeltaDictionary.